File: admin-portal/src/config/roles.ts
// admin-portal/src/config/roles.ts
// Centralized role configuration - single source of truth for frontend

export const USER_ROLES = {
  ADMIN: 'admin',
  OFFICE_MANAGER: 'office_manager',
  LAWYER: 'lawyer',
  PSYCHOLOGIST: 'psychologist',
  RECEPTIONIST: 'receptionist',
  EVENT_COORDINATOR: 'event_coordinator',
  CLIENT: 'client',
} as const;

export type UserRole = typeof USER_ROLES[keyof typeof USER_ROLES];

export const ROLE_DISPLAY_NAMES = {
  [USER_ROLES.ADMIN]: 'Administrador',
  [USER_ROLES.OFFICE_MANAGER]: 'Gerente de Oficina',
  [USER_ROLES.LAWYER]: 'Abogado/a',
  [USER_ROLES.PSYCHOLOGIST]: 'Psicólogo/a',
  [USER_ROLES.RECEPTIONIST]: 'Recepcionista',
  [USER_ROLES.EVENT_COORDINATOR]: 'Coordinador/a de Eventos',
  [USER_ROLES.CLIENT]: 'Cliente',
} as const;

export const ROLE_OPTIONS = Object.entries(ROLE_DISPLAY_NAMES).map(([value, label]) => ({
  value,
  label,
}));

// Helper functions
export const getRoleDisplayName = (role: string): string => {
  return ROLE_DISPLAY_NAMES[role as UserRole] || role;
};

export const isAdminRole = (role: string): boolean => {
  return role === USER_ROLES.ADMIN;
};

export const isManagementRole = (role: string): boolean => {
  return role === USER_ROLES.ADMIN || role === USER_ROLES.OFFICE_MANAGER;
};

export const isProfessionalRole = (role: string): boolean => {
  return role === USER_ROLES.LAWYER || role === USER_ROLES.PSYCHOLOGIST;
};

export const isStaffRole = (role: string): boolean => {
  const staffRoles: UserRole[] = [
    USER_ROLES.LAWYER,
    USER_ROLES.PSYCHOLOGIST,
    USER_ROLES.RECEPTIONIST,
    USER_ROLES.EVENT_COORDINATOR,
  ];
  return staffRoles.includes(role as UserRole);
};

// Additional helper functions for backward compatibility
export const getRolesForCaseCategory = (category: string): UserRole[] => {
  // Return all staff roles for any category - can be customized later
  return [
    USER_ROLES.LAWYER,
    USER_ROLES.PSYCHOLOGIST,
    USER_ROLES.RECEPTIONIST,
    USER_ROLES.EVENT_COORDINATOR,
  ];
};

export type StaffRoleKey = UserRole;

// Additional exports for backward compatibility
export const STAFF_ROLES = {
  [USER_ROLES.ADMIN]: {
    key: USER_ROLES.ADMIN,
    displayName: ROLE_DISPLAY_NAMES[USER_ROLES.ADMIN],
    department: 'Administration',
  },
  [USER_ROLES.OFFICE_MANAGER]: {
    key: USER_ROLES.OFFICE_MANAGER,
    displayName: ROLE_DISPLAY_NAMES[USER_ROLES.OFFICE_MANAGER],
    department: 'Management',
  },
  [USER_ROLES.LAWYER]: {
    key: USER_ROLES.LAWYER,
    displayName: ROLE_DISPLAY_NAMES[USER_ROLES.LAWYER],
    department: 'Legal',
  },
  [USER_ROLES.PSYCHOLOGIST]: {
    key: USER_ROLES.PSYCHOLOGIST,
    displayName: ROLE_DISPLAY_NAMES[USER_ROLES.PSYCHOLOGIST],
    department: 'Psychology',
  },
  [USER_ROLES.RECEPTIONIST]: {
    key: USER_ROLES.RECEPTIONIST,
    displayName: ROLE_DISPLAY_NAMES[USER_ROLES.RECEPTIONIST],
    department: 'Administration',
  },
  [USER_ROLES.EVENT_COORDINATOR]: {
    key: USER_ROLES.EVENT_COORDINATOR,
    displayName: ROLE_DISPLAY_NAMES[USER_ROLES.EVENT_COORDINATOR],
    department: 'Events',
  },
};

export const PERMISSIONS = {
  [USER_ROLES.ADMIN]: ['*'], // Admin has all permissions
  [USER_ROLES.OFFICE_MANAGER]: ['cases.read', 'cases.write', 'appointments.read', 'appointments.write', 'users.read', 'manage_offices'],
  [USER_ROLES.LAWYER]: ['cases.read', 'cases.write', 'appointments.read', 'appointments.write', 'documents.read'],
  [USER_ROLES.PSYCHOLOGIST]: ['cases.read', 'cases.write', 'appointments.read', 'appointments.write', 'documents.read'],
  [USER_ROLES.RECEPTIONIST]: ['cases.read', 'appointments.read', 'appointments.write', 'users.read'],
  [USER_ROLES.EVENT_COORDINATOR]: ['events.read', 'events.write'],
  [USER_ROLES.CLIENT]: ['cases.read', 'appointments.read'],
};

// Document types and permissions
export const CASE_DOCUMENT_TYPES = {
  LEGAL: 'legal',
  PSYCHOLOGICAL: 'psychological',
  GENERAL: 'general',
  MEDICAL: 'medical',
  FINANCIAL: 'financial',
} as const;

export const CASE_DOCUMENT_PERMISSIONS = {
  [USER_ROLES.ADMIN]: Object.values(CASE_DOCUMENT_TYPES),
  [USER_ROLES.LAWYER]: [CASE_DOCUMENT_TYPES.LEGAL, CASE_DOCUMENT_TYPES.GENERAL],
  [USER_ROLES.PSYCHOLOGIST]: [CASE_DOCUMENT_TYPES.PSYCHOLOGICAL, CASE_DOCUMENT_TYPES.GENERAL],
  [USER_ROLES.RECEPTIONIST]: [CASE_DOCUMENT_TYPES.GENERAL],
  [USER_ROLES.OFFICE_MANAGER]: Object.values(CASE_DOCUMENT_TYPES),
  [USER_ROLES.EVENT_COORDINATOR]: [CASE_DOCUMENT_TYPES.GENERAL],
  [USER_ROLES.CLIENT]: [],
};

// Helper functions
export const getAllRoles = (): Array<{ value: string; label: string }> => {
  return ROLE_OPTIONS;
};

export const isValidRole = (role: string): boolean => {
  return Object.values(USER_ROLES).includes(role as UserRole);
};

export const requiresOffice = (role: string): boolean => {
  return [
    USER_ROLES.OFFICE_MANAGER,
    USER_ROLES.LAWYER,
    USER_ROLES.PSYCHOLOGIST,
    USER_ROLES.RECEPTIONIST,
    USER_ROLES.EVENT_COORDINATOR,
  ].includes(role as any);
};

export const canViewDocumentType = (userRole: string, documentType: string): boolean => {
  const allowedTypes = (CASE_DOCUMENT_PERMISSIONS as any)[userRole] || [];
  return allowedTypes.includes(documentType);
};

export const getNavigationItemsForRole = (role: string) => {
  const baseItems = [
    { key: 'dashboard', label: 'Dashboard', path: '/', icon: 'DashboardOutlined' },
    { key: 'cases', label: 'Cases', path: '/app/cases', icon: 'FileTextOutlined' },
    { key: 'appointments', label: 'Appointments', path: '/app/appointments', icon: 'CalendarOutlined' },
  ];

  if (isAdminRole(role)) {
    return [
      ...baseItems,
      { key: 'users', label: 'Users', path: '/app/users', icon: 'UserOutlined' },
      { key: 'offices', label: 'Offices', path: '/app/offices', icon: 'BankOutlined' },
      { key: 'records', label: 'Records', path: '/app/records', icon: 'FileOutlined' },
      { key: 'reports', label: 'Reports', path: '/app/reports', icon: 'BarChartOutlined' },
    ];
  }

  if (isManagementRole(role)) {
    return [
      ...baseItems,
      { key: 'users', label: 'Users', path: '/app/users', icon: 'UserOutlined' },
      { key: 'records', label: 'Records', path: '/app/records', icon: 'FileOutlined' },
      { key: 'reports', label: 'Reports', path: '/app/reports', icon: 'BarChartOutlined' },
    ];
  }

  return baseItems;
};

export const getRoleDefinition = (role: string) => {
  return (STAFF_ROLES as any)[role] || { key: role, displayName: role, department: 'Unknown' };
};

export const getDashboardWidgetsForRole = (role: string) => {
  const baseWidgets = ['cases', 'appointments'];
  
  if (isAdminRole(role)) {
    return [...baseWidgets, 'users', 'system_health', 'reports'];
  }
  
  if (isManagementRole(role)) {
    return [...baseWidgets, 'users', 'reports'];
  }
  
  return baseWidgets;
};

File: api/config/roles.go
package config

import (
	"fmt"
	"strings"
)

// StaffRole represents a staff role in the CAF system
type StaffRole struct {
	Key         string `json:"key"`
	SpanishName string `json:"spanish_name"`
	EnglishName string `json:"english_name"`
	Department  string `json:"department"`
	Description string `json:"description"`
}

// Role constants - Single source of truth for all role keys
const (
	RoleAdmin            = "admin"
	RoleOfficeManager    = "office_manager"
	RoleLawyer           = "lawyer"
	RolePsychologist     = "psychologist"
	RoleReceptionist     = "receptionist"
	RoleEventCoordinator = "event_coordinator"
)

// STAFF_ROLES is the authoritative list of all valid staff roles
// This is the single source of truth for role definitions
var STAFF_ROLES = map[string]StaffRole{
	RoleAdmin: {
		Key:         RoleAdmin,
		SpanishName: "Administrador",
		EnglishName: "Administrator",
		Department:  "Administration",
		Description: "Full system access and management",
	},
	RoleOfficeManager: {
		Key:         RoleOfficeManager,
		SpanishName: "Gerente de Oficina",
		EnglishName: "Office Manager",
		Department:  "Management",
		Description: "Office-level management and oversight",
	},
	RoleLawyer: {
		Key:         RoleLawyer,
		SpanishName: "Abogado/a",
		EnglishName: "Lawyer",
		Department:  "Legal",
		Description: "Legal case management and documentation",
	},
	RolePsychologist: {
		Key:         RolePsychologist,
		SpanishName: "Psicólogo/a",
		EnglishName: "Psychologist",
		Department:  "Psychology",
		Description: "Psychological assessment and counseling",
	},
	RoleReceptionist: {
		Key:         RoleReceptionist,
		SpanishName: "Recepcionista",
		EnglishName: "Receptionist",
		Department:  "Administration",
		Description: "Front desk and appointment management",
	},
	RoleEventCoordinator: {
		Key:         RoleEventCoordinator,
		SpanishName: "Coordinador/a de Eventos",
		EnglishName: "Event Coordinator",
		Department:  "Events",
		Description: "Event planning and coordination",
	},
}

// VALID_ROLES contains all valid role keys for validation
var VALID_ROLES = []string{
	RoleAdmin,
	RoleOfficeManager,
	RoleLawyer,
	RolePsychologist,
	RoleReceptionist,
	RoleEventCoordinator,
}

// Role validation functions

// IsValidRole checks if a role key is valid
func IsValidRole(role string) bool {
	_, exists := STAFF_ROLES[role]
	return exists
}

// ValidateRole validates a role and returns an error if invalid
func ValidateRole(role string) error {
	if !IsValidRole(role) {
		return fmt.Errorf("invalid role '%s'. Valid roles are: %s", role, strings.Join(VALID_ROLES, ", "))
	}
	return nil
}

// GetRoleInfo returns role information for a given role key
func GetRoleInfo(role string) (StaffRole, error) {
	roleInfo, exists := STAFF_ROLES[role]
	if !exists {
		return StaffRole{}, fmt.Errorf("role '%s' not found", role)
	}
	return roleInfo, nil
}

// GetAllRoles returns all available roles
func GetAllRoles() map[string]StaffRole {
	return STAFF_ROLES
}

// GetRolesByDepartment returns all roles in a specific department
func GetRolesByDepartment(department string) map[string]StaffRole {
	roles := make(map[string]StaffRole)
	for key, role := range STAFF_ROLES {
		if role.Department == department {
			roles[key] = role
		}
	}
	return roles
}

// Permission checking functions

// CanAccessAllOffices checks if a role can access data from all offices
func CanAccessAllOffices(role string) bool {
	return role == RoleAdmin
}

// CanManageUsers checks if a role can manage users
func CanManageUsers(role string) bool {
	return role == RoleAdmin
}

// CanManageOffices checks if a role can manage offices
func CanManageOffices(role string) bool {
	return role == RoleAdmin
}

// CanManageFiles checks if a role can manage files
func CanManageFiles(role string) bool {
	return role == RoleAdmin
}

// CanManageWebContent checks if a role can manage web content
func CanManageWebContent(role string) bool {
	return role == RoleAdmin || role == RoleOfficeManager || role == RoleEventCoordinator
}

// CanViewLegalDocuments checks if a role can view legal documents
func CanViewLegalDocuments(role string) bool {
	return role == RoleAdmin || role == RoleLawyer
}

// CanViewPsychologicalDocuments checks if a role can view psychological documents
func CanViewPsychologicalDocuments(role string) bool {
	return role == RoleAdmin || role == RolePsychologist
}

// CanViewCases checks if a role can view cases
func CanViewCases(role string) bool {
	return role == RoleAdmin || role == RoleOfficeManager || role == RoleLawyer || role == RolePsychologist || role == RoleReceptionist
}

// CanViewAppointments checks if a role can view appointments
func CanViewAppointments(role string) bool {
	return role == RoleAdmin || role == RoleOfficeManager || role == RoleLawyer || role == RolePsychologist || role == RoleReceptionist
}

// CanViewEvents checks if a role can view events
func CanViewEvents(role string) bool {
	return role == RoleAdmin || role == RoleEventCoordinator
}

// IsProfessionalRole checks if a role is a professional service role
func IsProfessionalRole(role string) bool {
	return role == RoleLawyer || role == RolePsychologist
}

// IsAdministrativeRole checks if a role is an administrative role
func IsAdministrativeRole(role string) bool {
	return role == RoleAdmin || role == RoleOfficeManager || role == RoleReceptionist
}

// IsManagementRole checks if a role is a management role
func IsManagementRole(role string) bool {
	return role == RoleAdmin || role == RoleOfficeManager
}

// GetRoleHierarchyLevel returns the hierarchy level of a role (lower number = higher access)
func GetRoleHierarchyLevel(role string) int {
	switch role {
	case RoleAdmin:
		return 1
	case RoleOfficeManager:
		return 2
	case RoleLawyer, RolePsychologist:
		return 3
	case RoleReceptionist:
		return 4
	case RoleEventCoordinator:
		return 5
	default:
		return 999 // Unknown role
	}
}

// HasHigherOrEqualAccess checks if role1 has higher or equal access than role2
func HasHigherOrEqualAccess(role1, role2 string) bool {
	return GetRoleHierarchyLevel(role1) <= GetRoleHierarchyLevel(role2)
}

// Role validation middleware helper
func ValidateRoleMiddleware(role string) (bool, string) {
	if !IsValidRole(role) {
		return false, fmt.Sprintf("Invalid role '%s'. Valid roles: %s", role, strings.Join(VALID_ROLES, ", "))
	}
	return true, ""
}
